src/solver/solver.cxx

```diff
@@ -1357,6 +1383,140 @@ Field3D Solver::globalIndex(int localStart) {
  * Running user-supplied functions
  **************************************************************************/
 
+int Solver::run_rhs_se(BoutReal t, bool linear) {
+  int status;
+
+  Timer timer("rhs");
+
+  if (first_rhs_call) {
+    // Ensure that nonlinear terms are calculated on first call
+    linear = false;
+    first_rhs_call = false;
+  }
+
+  pre_rhs(t);
+  status = model->runRHS_se(t, linear);
+  post_rhs(t);
+
+  // If using Method of Manufactured Solutions
+  add_mms_sources(t);
+
+  rhs_ncalls_se++;
+  return status;
+}
+
+int Solver::run_rhs_si(BoutReal t, bool linear) {
+  int status;
+
+  Timer timer("rhs");
+
+  if (first_rhs_call) {
+    // Ensure that nonlinear terms are calculated on first call
+    linear = false;
+    first_rhs_call = false;
+  }
+
+  pre_rhs(t);
+  status = model->runRHS_si(t, linear);
+  post_rhs(t);
+
+  // If using Method of Manufactured Solutions
+  add_mms_sources(t);
+
+  rhs_ncalls_si++;
+  return status;
+}
+
+int Solver::run_rhs_fe(BoutReal t, bool linear) {
+  int status;
+
+  Timer timer("rhs");
+
+  if (first_rhs_call) {
+    // Ensure that nonlinear terms are calculated on first call
+    linear = false;
+    first_rhs_call = false;
+  }
+
+  pre_rhs(t);
+  status = model->runRHS_fe(t, linear);
+  post_rhs(t);
+
+  // If using Method of Manufactured Solutions
+  add_mms_sources(t);
+
+  rhs_ncalls_fe++;
+  return status;
+}
+
+int Solver::run_rhs_fi(BoutReal t, bool linear) {
+  int status;
+
+  Timer timer("rhs");
+
+  if (first_rhs_call) {
+    // Ensure that nonlinear terms are calculated on first call
+    linear = false;
+    first_rhs_call = false;
+  }
```

Comment:
The inner solver should use a different flag than the outer solver.  I don't recall if there was a `inner_first_rhs_call` flag in the class, but if not then I recommend that it be added (and used here)